### PR TITLE
Ensure single resource custom editors move instead of reopen

### LINKED ISSUE #122363: Opening a dirty custom editor to the side that cannot open more than once is weird
Testing #122030

I have a Luna paint image opened and dirty and select "Open to side" from the explorer. I get this dialog:

image

I would not expect getting asked to save for this action.

### PR DESCRIPTION
**Pull request by lramos15**
This PR fixes #122363

I'm not sure if this is the best way to do it, but I felt this logic shouldn't live in the override service as we don't want the override service doing weird things to hook into certain open cases we just want it to take a resource and return an input. So here I trigger a move when opening to the side if the editor cannot be opened in multiple groups.

### CODE REVIEW

**Review by bpasero — CHANGES_REQUESTED** [review]
I don't think asking users of the `SIDE_GROUP` to implement this properly is the right direction, there could be a ton of more places where we open to the side that is not the explorer and they would all run into this issue.

If the editor service allows to open an editor input to the side, imho it needs to deal with inputs that must move and cannot be copied internally.

**Comment by lramos15** [comment]
@bpasero There is currently only one consumer (from a quick check through, don't quote me on that), but I agree in the future that might not be the case. I have moved the logic into the editor service which allows us to keep `findSideBySideGroup` private.

File: src/vs/workbench/contrib/files/browser/fileCommands.ts [hunk 1]
```diff
@@ -144,7 +144,7 @@ KeybindingsRegistry.registerCommandAndKeybindingRule({
 
 				return await fileService.resolve(resource);
 			}));
-			const files = items.filter(i => !i.isDirectory);
+			let files = items.filter(i => !i.isDirectory);
```
Comment (bpasero) [inline src/vs/workbench/contrib/files/browser/fileCommands.ts:147]: Change is not needed.

File: src/vs/workbench/services/editor/browser/editorService.ts [hunk 2]
```diff
@@ -784,7 +784,7 @@ export class EditorService extends Disposable implements EditorServiceImpl {
 	async openEditors(editors: Array<IEditorInputWithOptions | IResourceEditorInputType>, group?: OpenInEditorGroup): Promise<IEditorPane[]> {
 
 		// Convert to typed editors and options
-		const typedEditors: IEditorInputWithOptions[] = editors.map(editor => {
+		let typedEditors: IEditorInputWithOptions[] = editors.map(editor => {
```
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:787]: Why does this only apply to `openEditors` what about `openEditor` method?

File: src/vs/workbench/services/editor/browser/editorService.ts [hunk 3]
```diff
@@ -798,7 +798,13 @@ export class EditorService extends Disposable implements EditorServiceImpl {
 		// Find target groups to open
 		const mapGroupToEditorsCandidates = new Map<IEditorGroup, IEditorInputWithOptions[]>();
 		if (group === SIDE_GROUP) {
+			// Custom editors which are open and only support one editor per resource should just move
+			const singletonEditors = this.getEditors(EditorsOrder.SEQUENTIAL).filter(e => typedEditors.find(t => isEqual(e.editor.resource, t.editor.resource)) && !e.editor.canSplit());
```
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:802]: I think the proper way to figure out if an editor matches should be to use it's `matches()` method, or not?

**Review by bpasero — CHANGES_REQUESTED** [review]

File: src/vs/workbench/services/editor/browser/editorService.ts (hunk 2 above)
Comment (lramos15) [reply src/vs/workbench/services/editor/browser/editorService.ts:787]: openEditors is the only one that currently acts on side by side groups since the explorer is the only place where we currently allow open to the side and when you open to the side you can open multiple

File: src/vs/workbench/services/editor/browser/editorService.ts (hunk 3 above)
Comment (lramos15) [reply src/vs/workbench/services/editor/browser/editorService.ts:802]: This is a tough one.... typedEditors is a set of FileEditorInputs (in the case of open to the side) therefore they will never match the custom editor ones in the `matches` call. I could argue it would work if I put this below the override logic, but the override logic has the logic which closes the additional editor inputs if we only support a single per resource leading to the problem above. 

In both scenarios this bug still exists 
1. Open icon.png in Luna paint
2. Switch to the hex editor using reopen with
3. Open icon.png to the side
4. 🐛 The hex editor moves. I should've gotten Luna paint to the side.

I could investigate how to fix this now, but I believe it's a challenging and complex issue which is too risky this late on. This solution prevents data loss, does something somewhat sensible, and we never supported opening custom editors to the side before.

**Comment by bpasero** [comment]
@lramos15 I think this needs a different solution. Just by looking at clients of `SIDE_GROUP` there is a ton of users throughout all areas:

image

To give an example, if you ctrl+click on a problem in the problem view, we allow you to open the file of that problem to the side. All these calls will go through `openEditor` and not `openEditors`.

To make matters worse: nothing prevents someone to use the API and open an editor in another group even if that editor declares it cannot split, by just calling `group.openEditor(input)`.

If we truly need to prevent editors to open in multiple groups that do not support that, I think the solution needs to be on a very low level somewhere in the group maybe (which is ugly). 

**To summarise:**
* there are editors that can only be opened in a single editor group (custom editor mainly)
* we have a ton of API and clients that open editors either through either editor service or direct group usage (less)
* anytime an editor opens in a group it potentially conflicts with its `canSplit` capability

Would this make sense:
* anytime an editor is to open in a group, `editor.canSplit` is called
* if an editor can split, proceed as normals
* if an editor cannot split, check all other groups for a matching editor (`editor.matches`) and move it into the group instead of opening it

Maybe we should move this to May given the amount of impact this could have.

**Comment by lramos15** [comment]
@bpasero What you proposed should work. Although like you said, living in the group is a little messy, but if that's what necessary. As for moving to May. I think we have to even though the current issue causes data loss we might do more harm than good.

**Comment by lramos15** [comment]
Removing fromt he milestone since this won't make it
